### ISSUE TITLE
Hibernate Envers integration

### ISSUE THREAD

**Issue by mgonzaga1990** [hibernate/hibernate-reactive#394]
Hi,

I would like to ask if we have a way to properly integrate hibernate envers?
either the hibernate envers of quarkus-hibernate-envers?

I tried using quarkus-hibernate-envers but it is giving me 

```
Caused by: org.hibernate.envers.exception.AuditException: Unable to create revision because of non-active transaction
```

I tried looking at the documentation but I can't find envers or auditing or is it currently not supported?

Thanks

**Comment by DavideD** [comment]
Hi,
Envers is currently not supported with Hibernate Reactive.
We haven't looked into it yet.

**Comment by mgonzaga1990** [comment]
Thanks for the fast response @DavideD .. But are you gonna look in to it in the future?

**Comment by Sanne** [comment]
Hi @mgonzaga1990 , we're not planning to look at it in the near future. We might revisit this in a year or more as there's many other things that need priority, so I'll close this for now.

**Comment by indiealexh** [comment]
I would like to bring this topic back up seeing as its over 1 year on and a feature I would love!

**Comment by DavideD** [comment]
I'm afraid the situation hasn't changed

**Comment by SMakhrov** [comment]
@Sanne , could you please clarify situation? We really need this

**Comment by Sanne** [comment]
@SMakhrov hi - while I think this would be really nice to have, this particular feature request is not very high in the list of strategic things we'll need to work on ourselves so it's unlikely that me or @DavideD will be implementing this, but if you or others volunteer to try it and send a pull request we'd be happy to help with guidance, suggestions and reviews.

At this point I don't know how hard it is - it's probably not too much work as Envers builds on top of the model that ORM core provides, so I expect only the cases in which Envers interfaces directly with statement execution would need to be adapted.

I would suggest make a POC ?

**Comment by DyadkinMaxim** [comment]
@Sanne, please could you say where I can find requirements for PR in this issue?
My team is analysing the way of making audit via reactive Envers, so probably we can take part in contribution.

**Comment by Sanne** [comment]
Let's bear in mind that we're going to upgrade Hibernate Reactive to Hibernate ORM v6+ soon (it's currently still based on ORM v5), so I expect the Envers related changes would be very different.

It might be worth finishing the upgrade to 6 first and only then work on Envers integration?

**Comment by gavinking** [comment]
Agreed.

**Comment by DyadkinMaxim** [comment]
@Sanne could you say when Hibernate Reactive will be updated to v6?

**Comment by indiealexh** [comment]
New Year. New need for auditing in Reactive Hibernate (◕ω◕✿)

**Comment by anothergoodguy** [comment]
Hi Team,

It's a great feature to have envers integrated with hibernate reactive...

Could you please share if it's in the plans?

**Comment by Sanne** [comment]
Hi @anothergoodguy ,

unfortunately as we're a small team I don't think we can work on this ourselves soon. But if anyone wants to give it a shot we're certainly happy to help with PR reviews and suggestions.

The upgrade to Hibernate ORM 6 is nearly done: we expect to release the version 2.0.0.Final soon, in a week or two, so it could be a good time for anyone to give this a shot.

**Comment by anothergoodguy** [comment]
Hi @Sanne 

Thank you for the quick response... 
We are also a very small crowd sourced team trying to build health portals. I'll try to spread the word and see if I can get some volunteers.

**Comment by vsolopov** [comment]
Hi @Sanne, do you have any terms regarding that? Trying to use hibernate Envers in Quarkus 2.16.6.Final but do not see any data in audit tables :c Look like hibernate reactive still doesn't support it.

**Comment by indiealexh** [comment]
Looking back into this. I don't know if I will find the time, but I am motivated to have this feature set so I will do some exploration.

**Comment by anothergoodguy** [comment]
hi @indiealexh any ticket to help testing and share feedback?

**Comment by indiealexh** [comment]
> hi @indiealexh any ticket to help testing and share feedback?

Not yet, busy work week so didn't make time for it, but check back next week :)

**Comment by anothergoodguy** [comment]
hi @indiealexh any ticket to help testing and share feedback?

**Comment by mgazim** [comment]
Hi there! Any updates on this?

**Comment by DavideD** [comment]
I don't think anybody is working on this.

**Comment by indiealexh** [comment]
> I don't think anybody is working on this.

Yeah, still trying to make time to try and contribute to this.

**Comment by anothergoodguy** [comment]
@indiealexh please share any ticker or branch so that we can help in testing !

**Comment by adampoplawski** [comment]
Hello
Is there some open MR/brach? I would like to look how big effort can be and maybe contribute.

**Comment by mikaelboffsysmo** [comment]
Hi @indiealexh any updates?

**Comment by indiealexh** [comment]
So, I haven't found time to work on this properly, but I did create a work around for now that might be useful for other AND/OR be useful for creating the envers reactive.

I will work on extracting it to a sample repo and then post here, but it is very "Quarkus" heavy in case someone is trying to use it outside that context. Additionally it has a limitation that the primary key has to be generated server side as I couldn't find an elegant way to create the main and audit record in one transaction otherwise (but thats probably just my lack of knowledge in this area).

**Comment by thmasker** [comment]
I'm also really interested in this

**Comment by thmasker** [comment]
Hi @Sanne I'd like to contribute to this but have no clue where to start...

**Comment by Sanne** [comment]
@thmasker excellent. I'd suggest to start by setting up a new source code module within the repository, add the hibernate-envers dependency to it and create an extremely simple test.
It's most likely going to fail, then you look into the errors and try to figure out what to do about it. At this point it's possible that it gets too tricky for you to continue alone, but don't worry and try to explain here how it's going - no shame and no judgement, we try to be nice to new contributors :)
Even if you decide to give up at that point, sharing the project setup and the first tests would be a useful contribution for others.

Probably a good place to discuss things would be the zulip chat, see https://hibernate.org/community/ for an overview or connect directly on https://hibernate.zulipchat.com - most maintainers and contributors hang out there.

**Comment by thmasker** [comment]
So, brief update: just created the test repository [hibernate-envers-reactive](https://github.com/thmasker/hibernate-envers-reactive) and it already fails (so far, so good, ironically). Will start investigation

**Comment by thmasker** [comment]
Only test the project has throws:
```
java.lang.AssertionError: 1 expectation failed.
Expected status code <200> but was <500>.
	Suppressed: org.hibernate.envers.exception.AuditException: Unable to create revision because of non-active transaction
		at org.hibernate.envers.event.spi.BaseEnversEventListener.checkIfTransactionInProgress(BaseEnversEventListener.java:135)
		at org.hibernate.envers.event.spi.EnversPostInsertEventListenerImpl.onPostInsert(EnversPostInsertEventListenerImpl.java:32)
		at org.hibernate.event.service.internal.EventListenerGroupImpl.fireLazyEventOnEachListener(EventListenerGroupImpl.java:129)
		at org.hibernate.action.internal.EntityInsertAction.postInsert(EntityInsertAction.java:215)
		at org.hibernate.reactive.engine.impl.ReactiveEntityRegularInsertAction.lambda$reactiveExecute$1(ReactiveEntityRegularInsertAction.java:79)
		at java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:757)
		at java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:735)
		at java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2214)
		at java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:144)
		at org.hibernate.reactive.engine.impl.ReactiveEntityRegularInsertAction.lambda$reactiveExecute$2(ReactiveEntityRegularInsertAction.java:74)
		at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
		at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
		at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
		at io.vertx.core.Future.lambda$toCompletionStage$3(Future.java:600)
		at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
		at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
		at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
		at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:88)
		at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:32)
		at io.vertx.core.Promise.complete(Promise.java:66)
		at io.vertx.core.Promise.handle(Promise.java:51)
		at io.vertx.core.Promise.handle(Promise.java:29)
		at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
		at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
		at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
		at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
		at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
		at io.vertx.sqlclient.impl.TransactionImpl.lambda$wrap$0(TransactionImpl.java:72)
		at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
		at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:60)
		at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
		at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:166)
		at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
		at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
		at io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java)
		at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:836)
		at io.netty.util.concurrent.SingleThreadEventExecutor.execute0(SingleThreadEventExecutor.java:827)
		at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:817)
		at io.vertx.core.impl.EventLoopExecutor.execute(EventLoopExecutor.java:35)
		at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:314)
		at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:168)
		at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:57)
		at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
		at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
		at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
		at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
		at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
		at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:324)
		at io.vertx.pgclient.impl.PgSocketConnection.handleMessage(PgSocketConnection.java:114)
		at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:137)
		at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:342)
		at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
		at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:389)
		at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
		at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
		at io.vertx.pgclient.impl.codec.PgDecoder.fireCommandResponse(PgDecoder.java:52)
		at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:133)
		at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:248)
		at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:107)
		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:1570)
```
I must say, I'm already lost here. How should I approach this integration? New module? "Fixing" library? I guess a mix is also possible?

**Comment by thmasker** [comment]
@sebersole I attached the project I'm using in https://github.com/hibernate/hibernate-reactive/issues/394#issuecomment-3170655362

**Comment by thmasker** [comment]
Basically, the exception is thrown because `AbstractLogicalConnectionImplementor#getStatus` returns `TransactionStatus.NOT_ACTIVE`

**Comment by sebersole** [comment]
I don't understand why people start off with "complicated" tests ;)  If I can make a suggestion, start simple.  Write a test that uses JUST reactive and envers.

As for why envers thinks that there is no active transaction... well for me that comes back to my point above.  You are assuming that panache is properly propogating that to ORM/Envers.  Which maybe it is, I don't know - again, I'm not really familiar with the innards of reactive let alone panache.  But the point is that it is an unnecessary uncertainty.

> It looks like the session is closed as soon as the insertion is made, thus failing running the postInsert() and we need to keep it alive somehow

So here is the rub, after digging through your test.  Repository uses StatelessSession.  Envers does not work with StatelessSession.  It very explicitly relies on the events and callbacks associated with Session.

**Comment by DavideD** [comment]
> I must say, I'm already lost here. How should I approach this integration? New module? "Fixing" library? I guess a mix is also possible?

I'm not familiar with envers, but I think it would make sense to follow the same structure we already have in Hibernate ORM.
So, it should be a separate module. I would for Hibernate Reactive `main` branch and add the everse module to it.

Generally speaking, Hibernate Reactive follows what's already implemented in Hibernate ORM. The difference is that when a class in Hibernate ORM has to run a query on the database, we need to rewrite it so that's going to run the same query using our own class: [SqlClientConnection](https://github.com/hibernate/hibernate-reactive/blob/main/hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/impl/SqlClientConnection.java)

We don't use the JDBC Driver.

I tend not to work on week-ends, I will have a better look next week. But feel free to ping me for specific questions.

Thanks for looking into this.

**Comment by thmasker** [comment]
Hi @sebersole, following your advice I've updated the test project with a much simpler (I hope) approach. I'll continue investigation from there

**Comment by sebersole** [comment]
@thmasker You might even want to start even simpler with just hibernate-envers and StatelessSession.  We'd need to work out problems there first - Envers will definitely not work with Reactive + Repository until we iron out everything with StatelessSession proper.

**Comment by sebersole** [comment]
Sorry, I meant even simpler :)

Remove Reactive from the equation.  Remove VertX from the equation.  Just use ORM - `hibernate-core` and `hibernate-envers` (and maybe `hibernate-testing`) - open a StatelessSession and attempt to persist some objects.  You'll see, like I said, this has nothing to do with Reactive per se.  Envers is simply not defined to work with StatelessSession atm.  E.g. -

```java
final SessionFactory sessionFactory = entityManagerFactory().unwrap( SessionFactory.class );
sessionFactory.inStatelessSession( (ss) -> {
	ss.insert( new Thing( 1, "it" ) );
} );
```

Leads to the same exact situation you are seeing -

```console

java.lang.NullPointerException: Cannot invoke "org.hibernate.engine.spi.SessionImplementor.isTransactionInProgress()" because "session" is null

	at org.hibernate.envers.event.spi.BaseEnversEventListener.checkIfTransactionInProgress(BaseEnversEventListener.java:132)
	at org.hibernate.envers.event.spi.EnversPostInsertEventListenerImpl.onPostInsert(EnversPostInsertEventListenerImpl.java:32)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireLazyEventOnEachListener(EventListenerGroupImpl.java:129)
	at org.hibernate.internal.StatelessSessionImpl.firePostInsert(StatelessSessionImpl.java:623)
	...
```

So there are 2 things here -

1. Making Envers work with StatelessSession
2. Making Envers work with Reactive - which it may or may not, I don't know - you'll have to TIAS.

So this may be as "simple" as getting Envers to work with StatelessSession, but either way that's what you'll HAVE TO address to make envers work with Jakarta Data and StatelessSession.

**Comment by sebersole** [comment]
P.S. sent you a PR for a simplified test

**Comment by DavideD** [comment]
@thmasker Maybe I'm missing something, but what are you trying to achieve here?

If your goal is to make the stateless session work with Hibernate Envers and Hibernate ORM, you should fork the Hibernate ORM repository, send a PR and discuss things there or on the Hibernate ORM jira.

If you want to work on the integration between Hibernate Envers and Hibernate Reactive, we can continue here. 

Mixing the two things only makes your life harder and this issue comment section quite noisy.

**Comment by DavideD** [comment]
Maybe I'm wrong, and @sebersole will confirm or deny that, but I think he saw your test was using the stateless session and was commenting and it currently doesn't work with Envers. Not that you need to implement that feature to use Envers with Hibernate Reactive.

I haven't seen your original test, though.

**Comment by sebersole** [comment]
> If your goal is to make the stateless session work with Hibernate Envers and Hibernate ORM, you should fork the Hibernate ORM repository, send a PR and discuss things there or on the Hibernate ORM jira.


> in order to make Hibernate Envers to work with Hibernate Reeactive (my actual goal) I need first to make the stateless session work with Hibernate Envers and Hibernate ORM

Not quite what I said, but close.  As Davide says, this has nothing to do with Reactive and its probably best, for the moment, to just remove that from your thoughts here.

The problem is that you are trying to use StatelessSession (via Repository) with Envers which DoesNotWork.  Never has.  It's not a bug, but an enhancement that you want.  So we (ahem, you) have to get that working first.  That's the gist of the test I sent you in PR.  It only uses `hibernate-core` (ORM) and `hibernate-envers`.  You have to get that working first.

> Now, I lack lots of knowledge to understand what the correct behaviour should be. Is it ok that it is null?

Yes, I explained already why it is.


> Should we create another session if it is null (does that even make sense?)?

No.  And no.

**Comment by sebersole** [comment]
> > Now, I lack lots of knowledge to understand what the correct behaviour should be. Is it ok that it is null?
> 
> Yes, I explained already why it is.
> 

Well, let me clarify a bit... no, its not ok that we pass null there in a vaccuum.  We need to refactor the relationship between `Session`, `StatelessSession` and `EventSource`.  That's the underlying first part of the solution here.  My gut tells me Envers will have other hard dependencies on `Session` but we'll (ahem, you'll ;) have to deal with those as they come up.

**Comment by sebersole** [comment]
> I would follow the following steps to work on this:

@DavideD We really need Envers to be working with `StatelessSession` first.  Any other effort at the present is moot.

**Comment by DavideD** [comment]
> [@DavideD](https://github.com/DavideD) We really need Envers to be working with `StatelessSession` first. Any other effort at the present is moot.

Ah, ok. That part wasn't clear to me. Everything else I said about the reactive side should still apply.

**Comment by thmasker** [comment]
> We need to refactor the relationship between `Session`, `StatelessSession` and `EventSource`

Biggest problem I see for newcomers as me is how can we understand such a complex project to be able to know the best way to handle a refactor like this.

**Comment by sebersole** [comment]
Actually, I spent a little time looking at this today and it was not as bad as I thought it would be on the ORM side.  Basically I split `AbstractDatabaseOperationEvent` into a separate hierarchy which uses `SharedSessionContractImplementor` instead of `EventSource`.  That works totally fine, including with StatelessSessions.

However, the rub is still Envers - mainly how we integrated it into ORM in terms of transaction callbacks.  At the moment it leverages `ActionQueue` which is only available on `Session`, not `StatelessSession`.  We will need to redesign that.
